drivers/sensor/qdec_nrfx/qdec_nrfx.c

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sensor.h>
+#include <spinlock.h>
+
+#include <nrfx_qdec.h>
+#include <hal/nrf_gpio.h>
+
+#define SYS_LOG_DOMAIN "qdec_nrfx"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_QDEC_NRFX_LEVEL
+#include <logging/sys_log.h>
+
+
+#define FULL_ANGLE 360
+
+/* limit range to avoid overflow when converting steps to degrees */
+#define ACC_MAX (INT_MAX / FULL_ANGLE)
+#define ACC_MIN (INT_MIN / FULL_ANGLE)
+
+
+struct qdec_nrfx_data {
+	struct k_spinlock        lock;
+	s32_t                    acc;
+	sensor_trigger_handler_t data_ready_handler;
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+	u32_t                    pm_state;
+#endif
+};
+
+
+static struct qdec_nrfx_data qdec_nrfx_data;
+
+DEVICE_DECLARE(qdec_nrfx);
+
+
+static void accumulate(struct qdec_nrfx_data *data, int16_t acc)
+{
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	bool overflow = ((acc > 0) && (ACC_MAX - acc < data->acc)) ||
+			((acc < 0) && (ACC_MIN - acc > data->acc));
+
+	if (!overflow) {
+		data->acc += acc;
+	}
+
+	k_spin_unlock(&data->lock, key);
+}
+
+static int qdec_nrfx_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct qdec_nrfx_data *data = &qdec_nrfx_data;
+
+	int16_t acc;
+	int16_t accdbl;
+
+	ARG_UNUSED(dev);
+
+	SYS_LOG_DBG("");
+
+	if ((chan != SENSOR_CHAN_ALL) && (chan != SENSOR_CHAN_ROTATION)) {
+		return -ENOTSUP;
+	}
+
+	nrfx_qdec_accumulators_read(&acc, &accdbl);
+
+	accumulate(data, acc);
+
+	return 0;
+}
+
+static int qdec_nrfx_channel_get(struct device       *dev,
+				 enum sensor_channel  chan,
+				 struct sensor_value *val)
+{
+	struct qdec_nrfx_data *data = &qdec_nrfx_data;
+	k_spinlock_key_t key;
+	s32_t acc;
+
+	ARG_UNUSED(dev);
+
+	SYS_LOG_DBG("");
+
+	if (chan != SENSOR_CHAN_ROTATION) {
+		return -ENOTSUP;
+	}
+
+	key = k_spin_lock(&data->lock);
+	acc = data->acc;
+	data->acc = 0;
+	k_spin_unlock(&data->lock, key);
+
+	static_assert(CONFIG_QDEC_STEPS > 0, "only positive number valid");
+	static_assert(CONFIG_QDEC_STEPS <= 2148, "overflow possible");
```

Comment:
use `__ASSERT` instead

Reply:
why adding check that runs every time (when enabled) instead of build time check?
this values does not change in runtime

Reply:
You're right, my mistake.